Firestore/core/test/unit/remote/watch_change_test.cc

```diff
@@ -43,19 +43,31 @@ TEST(WatchChangeTest, CanCreateDocumentWatchChange) {
 TEST(WatchChangeTest, CanCreateExistenceFilterWatchChange) {
   {
     ExistenceFilter filter{7, /*bloom_filter=*/absl::nullopt};
-    ExistenceFilterWatchChange change{filter, 5};
-    EXPECT_EQ(change.filter().count(), 7);
-    EXPECT_EQ(change.filter().bloom_filter_parameters(), absl::nullopt);
+    ExistenceFilterWatchChange change{std::move(filter), 5};
     EXPECT_EQ(change.target_id(), 5);
+    EXPECT_EQ(change.filter().count(), 7);
+    EXPECT_EQ(change.filter().bloom_filter(), absl::nullopt);
   }
   {
-    BloomFilterParameters bloom_filter_parameters{{0x42, 0xFE}, 7, 33};
-    ExistenceFilter filter{7, bloom_filter_parameters};
+    nanopb::Message<google_firestore_v1_BloomFilter> bloom_filter;
+    bloom_filter->hash_count = 33;
+    bloom_filter->has_bits = true;
+    bloom_filter->bits.padding = 7;
+    bloom_filter->bits.bitmap =
+        nanopb::MakeBytesArray(std::vector<uint8_t>{0x42, 0xFE});
+    ExistenceFilter filter{7, std::move(bloom_filter)};
     ExistenceFilterWatchChange change{std::move(filter), 5};
-    EXPECT_EQ(change.filter().count(), 7);
-    EXPECT_EQ(change.filter().bloom_filter_parameters(),
-              bloom_filter_parameters);
+
     EXPECT_EQ(change.target_id(), 5);
+    EXPECT_EQ(change.filter().count(), 7);
+
+    nanopb::Message<google_firestore_v1_BloomFilter> bloom_filter_copy;
+    bloom_filter_copy->hash_count = 33;
+    bloom_filter_copy->has_bits = true;
+    bloom_filter_copy->bits.padding = 7;
+    bloom_filter_copy->bits.bitmap =
+        nanopb::MakeBytesArray(std::vector<uint8_t>{0x42, 0xFE});
+    EXPECT_TRUE(change.filter().bloom_filter().value() == bloom_filter_copy);
```

Comment:
`EXPECT_EQ(change.filter().bloom_filter().value(), bloom_filter_copy);` crashes the build.

Reply:
The problem was that the `operator==` for `google_firestore_v1_BloomFilter` was defined in the wrong place. It was defined in the `firebase::firestore::remote` namespace in `existence_filter.h`, but should have been defined in the `firebase::firestore` namespace (the same namespace in which `google_firestore_v1_BloomFilter` is defined). Being in the wrong namespace prevented it from being found by ADL (address-dependent lookup) at compile time.

It seemed like a good idea to move the `operator==` functions out of the `existence_filter.h` file. So I did that by adding a new `operators.h` file in the `nanopb` directory. I also added unit tests for it (which need a bit more work in a follow-up PR).

I also added `operator==` for `nanopb::Message`, which was missing. This allows comparing `nanopb::Message` objects directly with each other, and eventually calling the `operator==` overloads for the underlying protos.

With these changes, the `EXPECT_EQ` now works, and the code is much better as well, with proper design and unit test coverage.

See https://github.com/firebase/firebase-ios-sdk/pull/10981/commits/7c7cb617af064db9648252333043ccffc441ed29